File: sdk/python/tempmail_sdk/providers/vip_215.py
"""
vip.215.im：POST /api/temp-inbox 创建收件箱；收件通过 WebSocket message.new 推送
推送无正文时统一合成 text/html（synthetic-v1），与 Node/Go/Rust 对齐。
"""

import html
import json
import threading
import time
from typing import Dict, List, Optional
from urllib.parse import quote

import websocket

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "vip-215"
BASE = "https://vip.215.im"

USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/148.0.0.0 Safari/537.36 Edg/148.0.0.0"
)

HOME_HEADERS = {
    "User-Agent": USER_AGENT,
    "Accept": (
        "text/html,application/xhtml+xml,application/xml;q=0.9,"
        "image/avif,image/webp,image/apng,*/*;q=0.8"
    ),
    "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
    "Cache-Control": "no-cache",
    "DNT": "1",
    "Pragma": "no-cache",
    "Sec-CH-UA": '"Chromium";v="148", "Microsoft Edge";v="148", "Not/A)Brand";v="99"',
    "Sec-CH-UA-Mobile": "?0",
    "Sec-CH-UA-Platform": '"Windows"',
    "Sec-Fetch-Dest": "document",
    "Sec-Fetch-Mode": "navigate",
    "Sec-Fetch-Site": "same-origin",
    "Sec-Fetch-User": "?1",
    "Upgrade-Insecure-Requests": "1",
}

HEADERS = {
    "User-Agent": USER_AGENT,
    "Accept": "*/*",
    "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
    "Cache-Control": "no-cache",
    "Content-Type": "application/json",
    "DNT": "1",
    "Origin": BASE,
    "Pragma": "no-cache",
    "Referer": f"{BASE}/",
    "Sec-CH-UA": '"Chromium";v="148", "Microsoft Edge";v="148", "Not/A)Brand";v="99"',
    "Sec-CH-UA-Mobile": "?0",
    "Sec-CH-UA-Platform": '"Windows"',
    "Sec-Fetch-Dest": "empty",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Site": "same-origin",
    "X-Locale": "zh",
}

WS_EXTRA_HEADERS = [
    f"User-Agent: {USER_AGENT}",
    f"Origin: {BASE}",
]


def _cookie_header_from_response(resp) -> str:
    jar = getattr(resp, "cookies", None)
    if jar:
        return "; ".join(f"{c.name}={c.value}" for c in jar)
    raw = resp.headers.get("Set-Cookie") or ""
    parts = []
    for segment in raw.split(","):
        nv = segment.split(";")[0].strip()
        if nv and "=" in nv:
            parts.append(nv)
    return "; ".join(parts)


def _establish_session() -> str:
    resp = tm_http.get(f"{BASE}/", headers=HOME_HEADERS, timeout=15)
    resp.raise_for_status()
    cookie = _cookie_header_from_response(resp)
    if "yyds_homepage_bridge=" not in cookie or "yyds_homepage_device=" not in cookie:
        raise RuntimeError("vip-215: missing homepage cookies")
    return cookie


def _fetch_ws_ticket(jwt: str) -> str:
    resp = tm_http.get(
        f"{BASE}/v1/auth/ws-ticket",
        headers={**HEADERS, "Authorization": f"Bearer {jwt}"},
        timeout=15,
    )
    resp.raise_for_status()
    body = resp.json()
    if not body.get("success"):
        raise RuntimeError("vip-215: ws-ticket success=false")
    ticket = (body.get("data") or {}).get("ticket")
    if not ticket:
        raise RuntimeError("vip-215: missing ws ticket")
    return ticket


_SYNTHETIC_MARKER = "【tempmail-sdk|synthetic|vip-215|v1】"

_lock = threading.Lock()
_boxes: Dict[str, "_Vip215Box"] = {}


class _Vip215Box:
    __slots__ = ("emails", "by_id", "thread", "started", "recipient")

    def __init__(self, recipient: str) -> None:
        self.emails: List[Email] = []
        self.by_id: Dict[str, Email] = {}
        self.thread: Optional[threading.Thread] = None
        self.started = False
        self.recipient = recipient


def _sanitize_one_line(val) -> str:
    if val is None:
        return ""
    s = str(val).replace("\r\n", " ").replace("\r", " ").replace("\n", " ")
    return s.strip()


def _build_synthetic_bodies(recipient_email: str, data: dict) -> tuple[str, str]:
    eid = _sanitize_one_line(data.get("id"))
    subj = _sanitize_one_line(data.get("subject"))
    from_ = _sanitize_one_line(data.get("from"))
    to = _sanitize_one_line(recipient_email)
    date = _sanitize_one_line(data.get("date"))
    size = data.get("size")
    lines = [
        _SYNTHETIC_MARKER,
        f"id: {eid}",
        f"subject: {subj}",
        f"from: {from_}",
        f"to: {to}",
        f"date: {date}",
    ]
    if isinstance(size, (int, float)) and not isinstance(size, bool) and size >= 0:
        lines.append(f"size: {int(size)}")
    text = "\n".join(lines)

    pairs = [
        ("id", eid),
        ("subject", subj),
        ("from", from_),
        ("to", to),
        ("date", date),
    ]
    parts = []
    for k, v in pairs:
        parts.append(f"<dt>{html.escape(k)}</dt><dd>{html.escape(v)}</dd>")
    if isinstance(size, (int, float)) and not isinstance(size, bool) and size >= 0:
        sv = str(int(size))
        parts.append(f"<dt>size</dt><dd>{html.escape(sv)}</dd>")
    inner = "".join(parts)
    html_body = (
        '<div class="tempmail-sdk-synthetic" data-tempmail-sdk-format="synthetic-v1" '
        'data-channel="vip-215">'
        f'<dl class="tempmail-sdk-meta">{inner}</dl></div>'
    )
    return text, html_body


def _get_box(token: str, recipient: str) -> _Vip215Box:
    with _lock:
        b = _boxes.get(token)
        if b is None:
            b = _Vip215Box(recipient)
            _boxes[token] = b
        return b


def _ws_run(jwt: str, box: _Vip215Box) -> None:
    ws_ticket = _fetch_ws_ticket(jwt)
    url = f"wss://vip.215.im/v1/ws?token={quote(ws_ticket, safe='')}"

    def on_message(_ws, message: str) -> None:
        try:
            msg = json.loads(message)
        except (json.JSONDecodeError, TypeError):
            return
        if msg.get("type") != "message.new":
            return
        data = msg.get("data") or {}
        syn_text, syn_html = _build_synthetic_bodies(box.recipient, data)
        raw = {
            "id": data.get("id"),
            "from": data.get("from"),
            "subject": data.get("subject"),
            "date": data.get("date"),
            "to": box.recipient,
            "text": syn_text,
            "html": syn_html,
        }
        em = normalize_email(raw, box.recipient)
        if not em.id:
            return
        with _lock:
            if em.id in box.by_id:
                return
            box.by_id[em.id] = em
            box.emails.append(em)

    ws_app = websocket.WebSocketApp(
        url,
        header=WS_EXTRA_HEADERS,
        on_message=on_message,
    )
    ws_app.run_forever(ping_interval=60, ping_timeout=30)


def _ensure_ws(token: str, recipient: str) -> None:
    box = _get_box(token, recipient)
    with _lock:
        if box.started:
            return
        box.started = True
        t = threading.Thread(target=_ws_run, args=(token, box), daemon=True)
        box.thread = t
        t.start()
    # 给握手留一点时间，避免首帧竞态
    time.sleep(0.08)


def generate_email() -> EmailInfo:
    cookie = _establish_session()
    hdrs = {**HEADERS, "Cookie": cookie}
    resp = tm_http.post(
        f"{BASE}/api/temp-inbox",
        headers=hdrs,
        data=b"",
        timeout=15,
    )
    resp.raise_for_status()
    body = resp.json()
    if not body.get("success"):
        raise RuntimeError("vip-215: success=false")
    data = body.get("data") or {}
    address = data.get("address")
    tok = data.get("token")
    if not address or not tok:
        raise RuntimeError("vip-215: missing address or token")
    _ensure_ws(tok, address)
    return EmailInfo(
        channel=CHANNEL,
        email=address,
        _token=tok,
        created_at=data.get("createdAt"),
        expires_at=None,
    )


def get_emails(token: str, email: str) -> List[Email]:
    _ensure_ws(token, email)
    box = _get_box(token, email)
    with _lock:
        return list(box.emails)

File: sdk/ruby/lib/tempmail_sdk/providers/vip_215.rb
# frozen_string_literal: true

require "json"
require "uri"
require "cgi"

module TempmailSdk
  module Providers
    # vip.215.im 渠道实现
    # 流程：GET / 取首页 cookie → POST /api/temp-inbox 创建邮箱（得到 address + token）
    #       GET /v1/messages 拉取邮件列表（HTTP 优先）
    #       GET /v1/messages/{id} 补拉单封详情
    # token 为 JWT，用于后续 API 鉴权
    module Vip215
      CHANNEL = "vip-215"
      BASE = "https://vip.215.im"
      UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " \
           "(KHTML, like Gecko) Chrome/148.0.0.0 Safari/537.36 Edg/148.0.0.0"

      HOME_HEADERS = {
        "User-Agent" => UA,
        "Accept" => "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Accept-Language" => "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
        "Cache-Control" => "no-cache",
        "DNT" => "1",
        "Pragma" => "no-cache",
        "Sec-Fetch-Dest" => "document",
        "Sec-Fetch-Mode" => "navigate",
        "Sec-Fetch-Site" => "same-origin",
        "Upgrade-Insecure-Requests" => "1"
      }.freeze

      API_HEADERS = {
        "User-Agent" => UA,
        "Accept" => "*/*",
        "Accept-Language" => "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
        "Cache-Control" => "no-cache",
        "Content-Type" => "application/json",
        "DNT" => "1",
        "Origin" => BASE,
        "Pragma" => "no-cache",
        "Referer" => "#{BASE}/",
        "Sec-Fetch-Dest" => "empty",
        "Sec-Fetch-Mode" => "cors",
        "Sec-Fetch-Site" => "same-origin",
        "X-Locale" => "zh"
      }.freeze

      module_function

      # 合并 Set-Cookie 响应头到 cookie 字符串
      def merge_cookies(existing, set_cookie_lines)
        map = {}
        existing.to_s.split(";").each do |part|
          part = part.strip
          next unless part.include?("=")

          k, v = part.split("=", 2)
          map[k.strip] = v.to_s.strip unless k.strip.empty?
        end
        set_cookie_lines.each do |line|
          pair = line.split(";", 2).first.to_s.strip
          next unless pair.include?("=")

          k, v = pair.split("=", 2)
          map[k.strip] = v.to_s.strip unless k.strip.empty?
        end
        map.map { |k, v| "#{k}=#{v}" }.join("; ")
      end

      # 访问首页获取必要 cookie（yyds_homepage_bridge / yyds_homepage_device）
      def establish_session
        resp = Http.get(BASE + "/", headers: HOME_HEADERS, timeout: 15)
        resp.raise_for_status
        cookie = merge_cookies("", resp.set_cookies)
        raise "vip-215: 缺少首页 cookie" unless cookie.include?("yyds_homepage_bridge=") &&
                                                  cookie.include?("yyds_homepage_device=")

        cookie
      end

      # 创建 vip-215 临时邮箱
      # @return [EmailInfo]
      def generate_email
        cookie = establish_session
        resp = Http.post(
          BASE + "/api/temp-inbox",
          headers: API_HEADERS.merge("Cookie" => cookie),
          body: "",
          timeout: 15
        )
        resp.raise_for_status
        data = JSON.parse(resp.body)
        raise "vip-215: 创建失败" unless data["success"] && data["data"]

        address = data["data"]["address"].to_s.strip
        token = data["data"]["token"].to_s.strip
        raise "vip-215: 响应缺少 address 或 token" if address.empty? || token.empty?

        EmailInfo.new(channel: CHANNEL, email: address, token: token)
      end

      # 判断邮件条目是否已含真实正文
      def has_real_body?(row)
        %w[text body_text html body_html content body].any? do |key|
          v = row[key]
          v.is_a?(String) && !v.strip.empty?
        end
      end

      # 提取邮件 ID
      def extract_message_id(row)
        %w[id messageId message_id].each do |key|
          v = row[key]
          next if v.nil?
          return v.to_s.strip if v.is_a?(String) && !v.strip.empty?
          return v.to_i.to_s if v.is_a?(Numeric)
        end
        ""
      end

      # 补拉单封邮件详情
      def fetch_message_detail(token, id)
        mid = id.to_s.strip
        return nil if mid.empty?

        resp = Http.get(
          "#{BASE}/v1/messages/#{URI.encode_www_form_component(mid)}",
          headers: API_HEADERS.merge("Authorization" => "Bearer #{token}"),
          timeout: 15
        )
        return nil unless resp.status_code == 200

        parsed = JSON.parse(resp.body)
        return nil unless parsed["success"] && parsed["data"].is_a?(Hash)

        parsed["data"]
      rescue StandardError
        nil
      end

      # 获取 vip-215 邮件列表（HTTP GET /v1/messages）
      # @param token [String] JWT
      # @param email [String]
      # @return [Array<Email>]
      def get_emails(token, email)
        raise "vip-215: token 为空" if token.to_s.strip.empty?

        resp = Http.get(
          BASE + "/v1/messages",
          headers: API_HEADERS.merge("Authorization" => "Bearer #{token}"),
          timeout: 15
        )
        resp.raise_for_status
        parsed = JSON.parse(resp.body)
        raise "vip-215: messages success=false" unless parsed["success"]

        messages = parsed.dig("data", "messages") || []
        messages.map do |row|
          id = extract_message_id(row)
          unless has_real_body?(row) || id.empty?
            detail = fetch_message_detail(token, id)
            row = row.merge(detail) if detail
          end
          row["to"] ||= email
          Normalize.normalize_email(row, email)
        end
      end
    end
  end
end
